File: Week 6/Problem Set 6/sentimental-credit/credit.py
# TODO
from sys import exit


def main():
    # Prompt user for input
    str_number = input("Number: ")
    lenght = len(str_number)
    if lenght in [13, 15, 16]:
        number = int(str_number)
    else:
        print("INVALID")
        exit(1)

    # First digit and the two first digits
    f = int(str_number[0])
    s = int(str_number[0] + str_number[1])

    # Luhn's Algorithm
    if luhn(number) == False:
        print("INVALID")
        exit(1)

    # Checking the company
    if (lenght == 15) and (s in [34, 37]):
        print("AMEX")
    elif (lenght == 16) and (s in [51, 52, 53, 54, 55]):
        print("MASTERCARD")
    elif (lenght in [13, 16]) and (f == 4):
        print("VISA")
    else:
        print("INVALID")


def luhn(number):
    copy = number
    other = 0
    rest = 0

    while True:
        # Non-other digits
        rest += copy % 10
        copy //= 10

        # Other digits
        tmp = (copy % 10) * 2
        if tmp >= 10:
            tmp = (tmp % 10) + (tmp // 10)
        other += tmp
        copy //= 10

        # End loop
        if copy <= 0:
            break

    if ((other + rest) % 10) == 0:
        return True
    else:
        return False


main()

File: Week 1/Problem Set 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
//Variable necessary to get the number of digits
    int Digits = 1;
//Number prompted by the user
    long N;
    do
    {
        N = get_long("Credit card number: ");
    }
    while (N <= 0);

//Luhn's Algorithm
    int Ar1[8] = {0};
    int Ar2[8] = {0};
    int ArS[8] = {0};
    long K = N;
    long U = N;
//Starting with the number’s second-to-last digit
    for (int i = 0; i < 8; i++)
    {
        U  /= 10;
        Ar1[i] = U % 10 * 2;
        U /= 10;
    }
//Other numbers
    for (int w = 0; w < 8; w++)
    {
        Ar2[w] = K % 10;
        K /= 100;
    }
//Break into digits and sum them
    for (int y = 0; y < 8; y++)
    {
        ArS[y] = Ar1[y] % 10;
        Ar1[y] /= 10;
        ArS[y] += Ar1[y];
    }
//Sum of the array 1
    int T1 = 0;
    for (int z = 0; z < 8; z++)
    {
        T1 += ArS[z];
    }
//Sum of the array 2
    int T2 = 0;
    for (int c = 0; c < 8; c++)
    {
        T2 += Ar2[c];
    }
//Array 1 + Array 2
    int Total = T1 + T2;
    Total = Total % 10;

//Getting the first and second number
    while (N >= 100)
    {
        N /= 10;
        Digits++;
    }
    int S = N;

//Getting the first number
    while (N >= 10)
    {
        N /= 10;
        Digits++;
    }
    int F = N;

//Luhn's algorithm
    if (Total != 0)
    {
        printf("INVALID\n");
    }

//Starts with 4 and has 13 or 16 digits
    else if ((Digits == 13 || Digits == 16) && (F == 4))
    {
        printf("VISA\n");
    }
//Starts with 51,52,53,54 or 55 and has 16 digits
    else if ((Digits == 16) && (S >= 51 && S <= 55))
    {
        printf("MASTERCARD\n");
    }
//Starts with 34 or 37 and has 15 digit number
    else if ((Digits == 15) && (S == 34 || S == 37))
    {
        printf("AMEX\n");
    }
//Other invalid numbers
    else
    {
        printf("INVALID\n");
    }
}